netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -365,10 +437,22 @@ public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
        * lifetime and we only want to configure it once.
        */
       if (handlers != null) {
-        ctx.pipeline().addFirst(handlers);
+        for (ChannelHandler handler : handlers) {
+          ctx.pipeline().addBefore(ctx.name(), null, handler);
+        }
+        ChannelHandler handler0 = handlers[0];
+        ChannelHandlerContext handler0Ctx = ctx.pipeline().context(handlers[0]);
         handlers = null;
+        if (handler0Ctx != null) { // The handler may have removed itself immediately
```

Comment:
Oh, and your view of the channel pipeline before and after the `channelRegistered()`s is correct.